src/openstack_cloud/openstack_manager.py

```diff
@@ -0,0 +1,378 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for handling interactions with OpenStack."""
+import json
+import logging
+import secrets
+from dataclasses import dataclass
+from typing import Iterable, Literal, Optional
+
+import jinja2
+import keystoneauth1.exceptions.http
+import openstack
+import openstack.compute.v2.server
+import openstack.connection
+import openstack.exceptions
+import openstack.image.v2.image
+from openstack.exceptions import OpenStackCloudException
+from openstack.identity.v3.project import Project
+
+from charm_state import Arch, ProxyConfig, SSHDebugConnection, UnsupportedArchitectureError
+from errors import (
+    OpenstackImageBuildError,
+    OpenStackUnauthorizedError,
+    RunnerBinaryError,
+    SubprocessError,
+)
+from github_client import GithubClient
+from github_type import RunnerApplication
+from runner_type import GithubPath
+from utilities import execute_command, retry
+
+logger = logging.getLogger(__name__)
+
+IMAGE_PATH_TMPL = "jammy-server-cloudimg-{architecture}-compressed.img"
+IMAGE_NAME = "jammy"
+BUILD_OPENSTACK_IMAGE_SCRIPT_FILENAME = "scripts/build-openstack-image.sh"
+
+
+def _create_connection(cloud_config: dict[str, dict]) -> openstack.connection.Connection:
+    """Create a connection object.
+
+    This method should be called with a valid cloud_config. See def _validate_cloud_config.
+    Also, this method assumes that the clouds.yaml exists on ~/.config/openstack/clouds.yaml.
+    See charm_state.py _write_openstack_config_to_disk.
+
+    Args:
+        cloud_config: The configuration in clouds.yaml format to apply.
+
+    Raises:
+        InvalidConfigError: if the config has not all required information.
+
+    Returns:
+        An openstack.connection.Connection object.
+    """
+    clouds = list(cloud_config["clouds"].keys())
+    if len(clouds) > 1:
+        logger.warning("Multiple clouds defined in clouds.yaml. Using the first one to connect.")
+    cloud_name = clouds[0]
+
+    # api documents that keystoneauth1.exceptions.MissingRequiredOptions can be raised but
+    # I could not reproduce it. Therefore, no catch here.
+    return openstack.connect(cloud_name)
+
+
+def list_projects(cloud_config: dict[str, dict]) -> list[Project]:
+    """List all projects in the OpenStack cloud.
+
+    The purpose of the method is just to try out openstack integration and
+    it may be removed in the future.
+
+    It currently returns objects directly from the sdk,
+    which may not be ideal (mapping to domain objects may be preferable).
+
+    Returns:
+        A list of projects.
+    """
+    conn = _create_connection(cloud_config)
+    try:
+        projects = conn.list_projects()
+        logger.debug("OpenStack connection successful.")
+        logger.debug("Projects: %s", projects)
+        # pylint thinks this isn't an exception
```

Comment:
It might be good to point out which exception is causing the problem for pylint. For a reader without context, it might be `OpenStackUnauthorizedError` or `keystoneauth1.exceptions.http.Unauthorized`. I feel explicitly mention the exception would be a good idea.

I encountered this issue as well. It seems pylint does not think `keystoneauth1.exceptions.http.Unauthorized` is an exception. However, from the source code it seems to be a child class of `Exception`. I am unable to figure why pylint is throwing this error.

I can also include this change in the follow-up PR as well. Just want to confirm if your find points toward `keystoneauth1.exceptions.http.Unauthorized` being the problem as well.

Reply:
Added a comment, thank you! I've also added a connection.authorize() as well - could you have a look?